### PR TITLE
Refactor thunkify to return proper thunk abstraction

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132407

This is superior to lru_cache because (1) it's more explicit and (2) it
doesn't leak the original function after it's been forced.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -405,7 +423,7 @@ def wrap_with_proxy(e: object, proxy: _NestedProxys, constant: Optional[_NestedT
             assert isinstance(proxy, Proxy)
             # NB: eagerly set meta here, so that the numbering is in order
             set_meta(proxy, e)
-            set_proxy_slot(e, tracer, lambda: proxy)
+            set_proxy_slot(e, tracer, thunkify(lambda: proxy))
```
Comment (albanD) [inline torch/fx/experimental/proxy_tensor.py:426]: Does this really make a difference? You just change the lambda capturing proxy with your class holding onto it in .r ?

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -174,7 +174,25 @@ def has_proxy_slot(obj: Tensor, tracer: _ProxyTracer) -> bool:
...
+    f: Optional[Callable[[], R]]
+    r: Optional[R]
+
+    __slots__ = ['f', 'r']
+
+    def __init__(self, f: Callable[[], R]):
+        self.f = f
+        self.r = None
+
+    def force(self) -> R:
```
Comment (albanD) [inline torch/fx/experimental/proxy_tensor.py:187]: Why .force() vs `.__call__()` ?

File: torch/fx/experimental/proxy_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:426]: The interface for Thunk is different, it's not callable you call .force() on it

File: torch/fx/experimental/proxy_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:187]: I want to be able to easily able to grep force sites where I force the thunk

**Review by albanD — APPROVED** [review]
Change sound good.
I have no clue what's going on with the typing in this file so I will assume that lintrunner runs here and is accurate :D

**Review by albanD — APPROVED** [review]
Ok!

**Pull request closed without merging**